crates/sui-core/src/execution_cache/cached_version_map.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cmp::Ordering;
+use std::collections::VecDeque;
+
+use sui_types::base_types::SequenceNumber;
+
+/// CachedVersionMap is a map from version to value, with the additional contraints:
+/// - The key (SequenceNumber) must be monotonically increasing for each insert. If
+///   a key is inserted that is less than the previous key, it results in an assertion
+///   failure.
+/// - Similarly, only the item with the least key can be removed.
+/// - The intent of these constraints is to ensure that there are never gaps in the collection,
+///   so that membership in the map can be tested by comparing to both the highest and lowest
+///   (first and last) entries.
+#[derive(Debug)]
+pub struct CachedVersionMap<V> {
+    values: VecDeque<(SequenceNumber, V)>,
+}
+
+impl<V> Default for CachedVersionMap<V> {
+    fn default() -> Self {
+        Self {
+            values: VecDeque::new(),
+        }
+    }
+}
+
+impl<V> CachedVersionMap<V> {
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
+
+    pub fn insert(&mut self, version: SequenceNumber, value: V) {
+        if !self.values.is_empty() {
+            let back = self.values.back().unwrap().0;
+            assert!(
+                back < version,
+                "version must be monotonically increasing ({} < {})",
+                back,
+                version
+            );
+        }
+        self.values.push_back((version, value));
+    }
+
+    pub fn all_lt_or_eq_rev<'a>(
+        &'a self,
+        version: &'a SequenceNumber,
+    ) -> impl Iterator<Item = &'a (SequenceNumber, V)> {
+        self.values
+            .iter()
+            .rev()
+            .take_while(move |(v, _)| v <= version)
```

Comment:
I am not sure how this works. Could you come up with a better name for this function with some comments?
It seems to me this will return empty vec if version is larger than the last version in the queue?

Reply:
you're right, it doesn't work. adding a test.